pubsub/google/cloud/pubsub_v1/publisher/batch/thread.py

```diff
@@ -0,0 +1,236 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import threading
+import time
+
+from google.cloud.pubsub_v1 import types
+from google.cloud.pubsub_v1.publisher import exceptions
+from google.cloud.pubsub_v1.publisher import futures
+from google.cloud.pubsub_v1.publisher.batch import base
+
+
+class Batch(base.Batch):
+    """A batch of messages.
+
+    The batch is the internal group of messages which are either awaiting
+    publication or currently in-flight.
+
+    A batch is automatically created by the PublisherClient when the first
+    message to be published is received; subsequent messages are added to
+    that batch until the process of actual publishing _starts_.
+
+    Once this occurs, any new messages sent to :meth:`publish` open a new
+    batch.
+
+    If you are using this library, you most likely do not need to instantiate
+    batch objects directly; they will be created for you. If you want to
+    change the actual batching settings, see the ``batching`` argument on
+    :class:`~.pubsub_v1.PublisherClient`.
+
+    Any properties or methods on this class which are not defined in
+    :class:`~.pubsub_v1.publisher.batch.BaseBatch` should be considered
+    implementation details.
+
+    Args:
+        client (~.pubsub_v1.PublisherClient): The publisher client used to
+            create this batch.
+        topic (str): The topic. The format for this is
+            ``projects/{project}/topics/{topic}``.
+        settings (~.pubsub_v1.types.BatchSettings): The settings for batch
+            publishing. These should be considered immutable once the batch
+            has been opened.
+        autocommit (bool): Whether to autocommit the batch when the time
+            has elapsed. Defaults to True unless ``settings.max_latency`` is
+            inf.
+    """
+    def __init__(self, client, topic, settings, autocommit=True):
+        self._client = client
+
+        # These objects are all communicated between threads; ensure that
+        # any writes to them are atomic.
+        self._futures = []
+        self._messages = []
+        self._size = 0
+        self._settings = settings
+        self._status = self.Status.ACCEPTING_MESSAGES
+        self._topic = topic
+
+        # If max latency is specified, start a thread to monitor the batch and
+        # commit when the max latency is reached.
+        self._thread = None
+        if autocommit and self._settings.max_latency < float('inf'):
+            self._thread = threading.Thread(target=self.monitor)
```

Comment:
This spawns a non-daemon thread which will block the application from exiting until `monitor` ends. Is this acceptable? I'm thinking it is considering the low deadline and the desire not to drop pending messages on the floor when exiting.